### PR TITLE
Mode2 Pitch export for UST

### PR DESCRIPTION
**Pull request by shine5402**
Using the Ramer–Douglas–Peucker algorithm to simplify pitch data, and write it into ust.

### CODE REVIEW

**Comment by shine5402** [comment]
A few test confirmed that it can get the work done, but at some circumstance pitch will have some distortion, like they did in Mode1. I think this is caused by dividing track pitch into notes, and I have no idea how to solve this. But hey, it's sort of rare, so...

File: src/main/kotlin/io/Ust.kt [hunk 1]
```diff
@@ -330,9 +333,17 @@ object Ust {
 
             if (features.contains(Feature.ConvertPitch)) {
                 builder.appendLine("PBType=5")
-                val pitchString = makeMode1PitchDataString(pitchData?.notes?.get(index))
+                val pitchString = makeMode1PitchDataString(pitchDataMode1?.notes?.get(index))
                 builder.appendLine("PitchBend=$pitchString")
                 builder.appendLine("PBStart=0")
+
+                val currMode2Pitch = pitchDataMode2?.notes?.get(index)
```
Comment (sdercolin) [inline src/main/kotlin/io/Ust.kt:340]: I think `mode2Pitch` is enough? or please spell the name fully

File: src/main/kotlin/process/RDPSimplification.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+// As Utaformatix use Pair<Long, Double> to describe pitch point, so...
+typealias Point = Pair<Long, Double>
+
+private fun perpendicularDistance(pt: Point, lineStart: Point, lineEnd: Point): Double {
+    var dx = (lineEnd.first - lineStart.first).toDouble()
+    var dy = lineEnd.second - lineStart.second
+
+    // Normalize
+    val mag = kotlin.math.hypot(dx, dy)
+    if (mag > 0.0) { dx /= mag; dy /= mag }
```
Comment (sdercolin) [inline src/main/kotlin/process/RDPSimplification.kt:20]: We don't use `;` at all lol.. Please write in multiple lines.

File: src/main/kotlin/process/RDPSimplification.kt [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+/*
+* The Ramer–Douglas–Peucker algorithm is a line simplification algorithm
+* for reducing the number of points used to define its shape.
+*
+* Wikipedia: https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm
+* Implementation reference: https://rosettacode.org/wiki/Ramer-Douglas-Peucker_line_simplification
+* */
+
+// As Utaformatix use Pair<Long, Double> to describe pitch point, so...
+typealias Point = Pair<Long, Double>
```
Comment (sdercolin) [inline src/main/kotlin/process/RDPSimplification.kt:12]: Somewhere we use `Pair<Long, Double?>`
So please make it private

File: src/main/kotlin/process/RDPSimplification.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    val pvdot = dx * pvx + dy * pvy
+
+    // Scale line direction vector and subtract it from pv
+    val ax = pvx - pvdot * dx
+    val ay = pvy - pvdot * dy
+
+    return kotlin.math.hypot(ax, ay)
+}
+
+/** The RDP(Ramer–Douglas–Peucker) simplification algorithm for line based shape.
```
Comment (sdercolin) [inline src/main/kotlin/process/RDPSimplification.kt:34]: Very slight problem but JavaDoc should be like:
```
/**
* ...
* ...
*/
```

(don't write in the first line)

File: src/main/kotlin/process/RDPSimplification.kt [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    val end = pointList.size - 1
+    for (i in 1 until end) {
+        val d = perpendicularDistance(pointList[i], pointList[0], pointList[end])
+        if (d > dmax) { index = i; dmax = d }
+    }
+
+    // If max distance is greater than epsilon, recursively simplify
+    return if (dmax > epsilon) {
+        // val recResults1 = mutableListOf<Point>()
+        // val recResults2 = mutableListOf<Point>()
```
Comment (sdercolin) [inline src/main/kotlin/process/RDPSimplification.kt:55]: Forget to remove?

File: src/main/kotlin/process/RDPSimplification.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    return kotlin.math.hypot(ax, ay)
+}
+
+/** The RDP(Ramer–Douglas–Peucker) simplification algorithm for line based shape.
+ *  Slightly modified from rosettacode.org's implementation.
+ *  @param pointList Points that form the shape, connected by lines.
+ *  @param epsilon Defines how much the algorithm should simplify. Bigger value leads to more point to drop.
+ */
+fun simplifyShape(pointList: List<Point>, epsilon: Double): List<Point> {
+    if (pointList.size < 2)
```
Comment (sdercolin) [inline src/main/kotlin/process/RDPSimplification.kt:40]: Very slight but you can put it in one line:
```kotlin
if (pointList.size < 2) return
```

File: src/main/kotlin/process/RDPSimplification.kt [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+}
+
+/**
+ * Simplify given shape, which should later be described by [maxPointCount] (or less) points.
+ *
+ * Using [simplifyShape], which implements The RDP(Ramer–Douglas–Peucker) algorithm.
+ * */
+fun simplifyShapeTo(pointList: List<Point>, maxPointCount: Long): List<Point> {
+    if (pointList.size < maxPointCount)
```
Comment (sdercolin) [inline src/main/kotlin/process/RDPSimplification.kt:75]: ditto.

File: src/main/kotlin/process/Resample.kt [hunk 8]
```diff
@@ -19,8 +19,9 @@ fun List<Pair<Long, Double?>>.resampled(
     return result.toList()
 }
 
-// A shorthand for pitch represented in dot.
-// Its interpolateMethod simply copy the value from prev, or next if prev not exists.
+/** A shorthand for pitch represented in dot.
```
Comment (sdercolin) [inline src/main/kotlin/process/Resample.kt:22]: ditto.

**Comment by sdercolin** [comment]
LGTM

**Pull request merged**